fms_extras/models/sphinx.py

```diff
@@ -0,0 +1,476 @@
+import json
+import math
+import os
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Mapping, Optional, OrderedDict
+
+import torch
+import torch.nn as nn
+from fms import models
+from fms.distributed.strategy import (
+    DistributedStrategy,
+    NoOpStrategy,
+    TensorParallelStrategy,
+    UniformModelParallelStrategy,
+)
+from fms.modules.attention import MultiHeadAttention
+from fms.modules.embedding import WordEmbedding
+from fms.modules.feedforward import GatedLinearUnit
+from fms.modules.layernorm import LayerNormParameterized
+from fms.modules.positions import RotaryEmbedding
+from fms.utils import serialization
+from fms.utils.activation import str_to_activation
+from fms.utils.config import ModelConfig
+from fms.utils.tokenizers import _has_hf, get_tokenizer
+
+
+# params emb_dim heads layers lr
+#  7B    4096    32    32     3.0E-04
+# 13B    5120    40    40     3.0E-04
+# 33B    6656    52    60     1.5.E-04
+# 65B    8192    64    80     1.5.E-04
+
+
+@dataclass
+class SphinxConfig(ModelConfig):
+    src_vocab_size: int = 65024  # can be set by tokenizer
+    emb_dim: int = 4608
+    norm_eps: float = 1e-5
+    nheads: int = 36
+    kvheads: int = 4
+    nlayers: int = 46
+    pad_id: int = 2
+    hidden_grow_factor: float = 8 / 3
+    multiple_of: int = 1
+    activation_fn: str = "swish"
+    p_dropout: float = 0.1
+    max_expected_seq_len: int = 4096
+    ntk_scaling: bool = False
+
+
+class SphinxBlock(nn.Module):
```

Comment:
we should have some kind of comment explaining that this is an IBM model similar to llama and what the differences are (includes a bias and something else?).

Was there any consensus on re-naming this for open source? there exists a (multimodal) LLM with the same name: https://arxiv.org/abs/2311.07575

Reply:
Added comments regarding the differences

Reply:
name changed to Calico